jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py

```diff
@@ -177,3 +199,53 @@ def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
                     'return_embeddings', self.default_return_embeddings
                 ),
             )
+
+    @requests(on='/snapshot')
+    def snapshot(self, **kwargs):
+        # TODO argument with table name, database location
+        # create a duplicate of the table
+        # or send to another PSQL instance to avoid perf hit?
+        with self.handler as postgres_handler:
+            postgres_handler.snapshot()
+
+    def get_snapshot(self, shard_id: int, total_shards: int):
+        shards_to_get = self._vshards_to_get(shard_id, total_shards)
+
+        with self.handler as postgres_handler:
+            return postgres_handler.get_snapshot(shards_to_get)
+
+    def _vshards_to_get(self, shard_id, total_shards):
+        if shard_id > total_shards - 1:
+            raise ValueError('shard_id is 0-indexed out of range(total_shards)')
+        vshards = list(range(self.total_shards))
+        vshard_part = (
+            self.total_shards // total_shards
+        )  # nr of virtual shards given to one shard
+        vshard_remainder = vshard_part % total_shards
+        if shard_id == total_shards - 1:
+            shards_to_get = vshards[
+                shard_id
+                * vshard_part : ((shard_id + 1) * vshard_part + vshard_remainder)
+            ]
+        else:
+            shards_to_get = vshards[
+                shard_id * vshard_part : (shard_id + 1) * vshard_part
+            ]
+        return [str(shard_id) for shard_id in shards_to_get]
```

Comment:
This should be a `@staticmethod`, getting `self.total_shards` (however this will be renamed) as an input as well. Then it is super easy to heavily test it, since nobody is able to get, whether this function is right without tests :p

Reply:
Yes, good point. Done